blog/2020/05/invalidations.md

```diff
@@ -0,0 +1,784 @@
+
+## Fixing invalidations
+
+You may have noticed that two packages, `Example` and `Revise`, trigger far fewer invalidations that the rest of the packages in our analysis.
+`Example` is quite trivial, but `Revise` and its dependencies are quite large.
+How does it avoid this problem?
+First, Revise does not extend very many Base methods;
+most of its methods are for functions it "owns," and the same is true for its dependencies.
+Second, in the closing days of Julia 1.5's merge window,
+Revise (and Julia) underwent a process of tracking down invalidations and eliminating them;
+for comparison, on Julia 1.4, Revise triggers more than a 1000 non-unique invalidations.
+The success of this effort gives one hope that other packages too may one day have fewer invalidations.
+
+As stated above, there is reason to hope that most of the invalidations marked as "ambiguous" will be fixed by changes to Julia's compiler.
+Here our focus is on those marked "more specific," since those are cases where it is harder to imagine a generic fix.
+
+### Fixing type instabilities
+
+In engineering Julia and Revise to reduce invalidations, at least two cases were fixed by resolving type-instabilities.
+For example, one set of invalidations happened because `CodeTracking`, a dependency of Revise's, defines new methods for `Base.PkgId`.
+It turns out that this triggered an invalidation of `_tryrequire_from_serialized`, which is used to load packages.
+Fortunately, it turned out to be an easy fix: one section of `_tryrequire_from_serialized` had a passage
+
+```
+for M in mod::Vector{Any}
+    if PkgId(M) == modkey && module_build_id(M) === build_id
+        return M
+    end
+end
+```
+
+and since `M` had type `Any`, the compiler couldn't predict which version of `PkgId` would be called.
+It sufficed to add
+
+```
+    M = M::Module
+```
+
+immediately after the `for` statement to fix the problem.
+Not only does this fix the invalidation, but it lets the compiler generate better code.
+
+The other case was a call from `Pkg` of `keys` on an AbstractDict of unknown type
+(due to a caller's `@nospecialize` annotation).
+Replacing `keys(dct)` with `Base.KeySet(dct)` (which is the default return value of `keys`) eliminated a very consequential invalidation, one that triggered seconds-long latencies in the next `Pkg` command after loading Revise.
+The benefits of this change in Pkg's code went far beyond helping Revise; any package depending on the OrderedCollections package (which is a dependency of Revise and what actually triggered the invalidation) got the same benefit.
+With these and a few other relatively simple changes, loading Revise no longer forces Julia to recompile much of Pkg's code the next time you try to update packages.
+
+### Redirecting call chains
+
+Let's return to our FixedPointNumbers `reduce_empty` example above.
+A little prodding as done above reveals that this corresponds to the definition
+
+```julia-repl
```

Comment:
A refresher of the inserted method is helpful here (I had to scroll back up to grok this section)

```suggestion
```julia-repl
julia> tree = trees[end-1]
 insert reduce_empty(::typeof(Base.mul_prod), ::Type{F}) where F<:FixedPoint in FixedPointNumbers at /home/tim/.julia/packages/FixedPointNumbers/w2pxG/src/FixedPointNumbers.jl:225 invalidated:
    backedges: MethodInstance for reduce_empty(::Function, ::Type{T} where T) triggered MethodInstance for reduce_empty(::Base.BottomRF{typeof(max)}, ::Type{VersionNumber}) (136 children) more specific

```